probe-rs/src/probe/cmsisdap/commands/mod.rs

```diff
@@ -388,6 +391,24 @@ impl CmsisDapDevice {
             }
         }
     }
+
+    pub(super) fn connect_if_needed(&mut self) -> Result<(), DebugProbeError> {
+        match self {
+            CmsisDapDevice::V1 { .. } | CmsisDapDevice::V2 { .. } => {}
+            CmsisDapDevice::Tcp {
+                handle, address, ..
+            } => {
+                let mut buf = [0u8; 1];
+                if let Ok(0) = handle.borrow().peek(&mut buf) {
```

Comment:
Were you thinking:
```suggestion
                if let Err(_) = handle.borrow().peek(&mut buf) {
```

(possibly with a check that the error is a timeout?)

Reply:
Ah, I was not. I was making the assumption that the dettatch would close the connection and that would have propagated to the stream by this point (with the above being a check for that).

Reply:
Looking now at the dettach method that's completely bullshit. Pretty sure I must have been thinking that we could watch for a host disconnect LED event post flash operation, then use that to close the socket and trigger the reboot. That should make it agnostic over whether it's a flash or a flash then verify operation.

Reply:
https://github.com/chrisprice/probe-rs/blob/84d3b44b52ee9976b14e9afe9a302353c5190514/probe-rs/src/probe/cmsisdap/mod.rs#L885

Reply:
https://github.com/chrisprice/embassy-net-rp-self-debug/issues/3 https://github.com/chrisprice/embassy-net-rp-self-debug/issues/4

Reply:
> make it agnostic over whether it's a flash or a flash then verify operation

Hadn't thought of that spanner in the works, I suppose we could reset the chip in our handler for it? (just realised we get a disconnect event - chrisprice/embassy-net-rp-self-debug#5)

And if we close the socket chip-side, there'll be no timeout as probe-rs will see the socket shutdown immediately (but probably misbehave)